File: serveur/app/models/health_metric.py
# app/models/health_metric.py
#
# Santé / Health Connect V1 (2026-06-17). Type A user-scoped.
# Métriques passives génériques, vendor-agnostiques. Une row = une mesure.
# `type` UPPER_CASE (politique 11) : HEART_RATE / SLEEP / DISTANCE / ACTIVE_CALORIES.
# `unit` self-describing (bpm / min / m / km / kcal...) pour rester agnostique.
# Ancrage temporel : `date` (jour, requis, indexé) + `start_time` optionnel
# ("HH:MM") pour les mesures intraday (ex. échantillon de fréquence cardiaque).

from sqlalchemy import Column, Integer, String, Float, ForeignKey, DateTime
from app.database import Base
import uuid


class HealthMetric(Base):
    __tablename__ = "health_metrics"

    id = Column(Integer, primary_key=True, index=True)
    user_id = Column(Integer, ForeignKey("users.id", ondelete="CASCADE"), nullable=False, index=True)
    uuid = Column(String, nullable=False, unique=True, default=lambda: str(uuid.uuid4()))

    type = Column(String, nullable=False)               # UPPER_CASE : HEART_RATE | SLEEP | DISTANCE | ACTIVE_CALORIES
    value = Column(Float, nullable=False)
    unit = Column(String, nullable=False)               # bpm | min | m | km | kcal...
    date = Column(String, nullable=False, index=True)   # "YYYY-MM-DD" (convention projet)
    start_time = Column(String, nullable=True)          # "HH:MM" instant précis optionnel (mesures intraday)

    updated_at = Column(DateTime(timezone=True), nullable=True)

    def __repr__(self):
        return (
            f"HealthMetric(id={self.id}, user_id={self.user_id}, uuid='{self.uuid}', "
            f"type='{self.type}', value={self.value}, unit='{self.unit}', date='{self.date}')"
        )

File: appli-android/app/src/main/java/com/example/sportapp/core/data/model/HealthMetric.kt
package com.example.sportapp.core.data.model

import androidx.room.ColumnInfo
import androidx.room.Entity
import androidx.room.Index
import androidx.room.PrimaryKey
import com.example.sportapp.core.utils.CustomDateUtils.getNowISO8601

/**
 * Santé / Health Connect V1 (2026-06-17) — portage Android du modèle serveur
 * `health_metrics` (cf. serveur/app/models/health_metric.py). Type A user-scoped.
 *
 * Métriques passives génériques, vendor-agnostiques. Une row = une mesure.
 * `type` UPPER_CASE (politique 11) : HEART_RATE / SLEEP / DISTANCE / ACTIVE_CALORIES.
 * `unit` self-describing (bpm / min / m / km / kcal...). Ancrage temporel :
 * `date` (jour, requis, indexé) + `startTime` optionnel ("HH:MM") pour l'intraday.
 */
@Entity(
    tableName = "health_metrics",
    indices = [
        Index(value = ["uuid"], unique = true),
        Index(value = ["user_id"]),
        Index(value = ["date"]),
    ]
)
data class HealthMetric(
    @PrimaryKey
    @ColumnInfo(name = "uuid") val uuid: String,

    @ColumnInfo(name = "user_id") val userId: Int,

    val type: String,                                       // UPPER_CASE
    val value: Float,
    val unit: String,                                       // bpm | min | m | km | kcal...
    val date: String,                                       // "YYYY-MM-DD"
    @ColumnInfo(name = "start_time") val startTime: String? = null,  // "HH:MM" intraday optionnel

    @ColumnInfo(defaultValue = "0") val synced: Boolean = false,
    @ColumnInfo(defaultValue = "0") val pendingDeletion: Boolean = false,

    @ColumnInfo(name = "updated_at") val updatedAt: String? = getNowISO8601(),
)
